source/sliding_puzzle.py

```diff
@@ -126,21 +126,47 @@ def solved(self) -> bool:
                 i += 1
         return True
 
-    def _join(self, row: List[PuzzlePiece]) -> List[str]:
-        """Join together a row of puzzle pieces into a complete list of lines."""
-        res = []
-        for i in range(len(row[0].data)):
-            res.append('')
-            for piece in row:
-                res[i] += piece.data[i]
-
-        return res
-
     def draw(self) -> str:
         """Draw the puzzle in its current state."""
-        output = []
-        for row in self.rows:
-            output.extend(self._join(row))
+        nbr_separators = len(self.rows) - 1
+        piece_width = self.rows[0][0].width
+        puzzle_width = piece_width * len(self.rows[0]) + nbr_separators
+
+        def build_border(start: str, middle: str, end: str) -> str:
```

Comment:
Why did you opt to define these inside the method rather than having them as `self.build_border()`.

Reply:
there are used only once, having them in the draw method namespace seems to be the right place to me